### PR TITLE
alarm: rewrite alarm virtualization with better comments and simpler logic, add tests

### PR DESCRIPTION
**Pull request by alevy**
Relies on tock/tock#4201 to work.

This PR fixes two bugs in alarm virtualization, the second discovered when testing a fix for the first:

First, #466 attempted to fix how alarms were added to the global list of outstanding alarms, but missed a case. In particular, it's possible for only `cur` to overflow and still happen before the alarm being added (see inline comments added to `root_insert`.

This bug manifests when a client requests two alarms that are longer than `MAX_TICKS`, each of which gets implemented as a maximum length alarm that sets an additional alarm. This case can result in the longer of the two alarms being incorrectly moved to the back of the list.

While testing the fix, I discovered that alarm overflows aren't handled properly. This was due to poor interaction between a bug in the kernel (tock/tock#4201) and a complex virtualization logic that allowed new alarms to be added to the global list from alarm callbacks while iteration through the list was happening.

tock/tock#4201 allows for a simpler virtualization logic that defers callbacks until after checking for expiration.

This PR rewrites that the virtualization with that logic as well as adds extensive comments documenting the invariants assumed by virtualization. It also adds a test that checks this specific regression.

### CODE REVIEW

File: libtock/services/alarm.c [hunk 1]
```diff
@@ -145,27 +180,118 @@ static libtock_alarm_ticks_t* root_peek(void) {
...
+  // We know this upcall is associated with the head, so it's easier
+  // to deal with.
+  libtock_alarm_ticks_t* head = root_pop();
+
+  // Formally, we should be able to just add head to
+  // `tocall`, but let's just be defensive just in case there is an
+  // errant alarm.
+  assert(head != NULL && !is_within(scheduled, head->reference, head->dt));
+  head->next  = NULL;
+  head->prev  = tocall_last;
```
Comment (tyler-potyondy) [inline libtock/services/alarm.c:242]: Will this not always be `NULL`?

**Review by tyler-potyondy — CHANGES_REQUESTED** [review]
This is a major improvement over the existing virtualizer implementation! I appreciate the thorough explanations for the logic. 

With my first pass over this, it appears your stated logic in the comments matches what you have implemented. I will need to make another pass to confirm that the logic you describe in the comments is accurate and accounts for corner cases.

File: examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+1 second, while the third should fire after a clock overflow + 10ms +
+10 seconds.
+
+If the virtual alarm library is buggy, this test might fail by never
+firing the first alarm or firing it after the third alarm
+(specifically almost another clock overflow) because either alarm
+insertion or virtual alarm upcall handing misses it after the second
+alarm fires.
+
+# Example Output
```
[the remark this answers is not in the document]
Comment (alevy) [reply examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md:25]: I suppose? the numbers other than the first column are meaningless except for debugging maybe and they won't be the same across runs because the application boots at slightly different times.

The first column is just the expected order, and should be the same for all boards.

But yes, the nrf5x

File: examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md
[the remark this answers is not in the document]
Comment (alevy) [reply examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md:1]: Makes sense, but how about we do this in a different PR to avoid having too much unrelated stuff in this one. I would do it immediately after merging this.

File: libtock/services/alarm.c (hunk 1 above)
Comment (alevy) [reply libtock/services/alarm.c:242]: `tocall_last`? It will, yeah. I'm fine changing this to `NULL`, though stylistically I could go both ways. i.e. this matches a pattern that we always set `->prev` to whatever the previously last thing in the list was, as opposed to special-casing the head. On the other hand, maybe it's unnecessary cognitive burden to read through this way.

File: examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (alevy) [reply examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md:25]: Added a note, below

File: libtock/services/alarm.c (hunk 1 above)
Comment (alevy) [reply libtock/services/alarm.c]: Fixed

File: examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md
[the remark this answers is not in the document]
Comment (alevy) [reply examples/tests/multi_alarm_dont_ignore_previous_overflow/README.md:1]: Done

**Review by lschuermann — APPROVED** [review]
I'm not going to lie -- reviewing this gives me headaches. I don't see any red flags with these changes and I think I follow most of the logic, but I have not (and probably will not, in any reasonable amount of time) get to a point that I'm fully convinced the timer virtualization here is correct.

That being said, this fixes real bugs and adds test cases, so ... let's get this in!

**Comment by alevy** [comment]
@tyler-potyondy you are the decider now

**Review by bradjc — APPROVED** [review]
